perf/reader.go

```diff
@@ -374,6 +374,11 @@ func (pr *Reader) ReadInto(rec *Record) error {
 					pr.epollRings = append(pr.epollRings, ring)
 				}
 			}
+
+			if len(pr.epollRings) == 0 {
+				// Woken up but no data, deadline exceeded.
+				return fmt.Errorf("read: %w", os.ErrDeadlineExceeded)
```

Comment:
This return without checking errors.Is(err, os.ErrDeadlineExceeded) is invalid.
As in you patch you replace pr.epollEvents to one single event. In case there are 2 signaled files, we will read both of them. And when we will call _, err := pr.poller.Wait(events, pr.deadline) one more time - it will return second signaled file. But it was read. In this case we will return ErrDeadlineExceeded which was not triggered.

Reply:
I think @albe19029 is correct. From https://man7.org/linux/man-pages/man2/epoll_wait.2.html :
> If more than maxevents file descriptors are ready when epoll_wait() is called, then successive epoll_wait() calls will round robin through the set of ready file descriptors

Reply:
Ah interesting, basically a spurious wakeup? I'll have a think!

Reply:
I pushed a fixed version, thanks for pointing this out. Important caveat in the current implementation (which applies to `ringbuf` as well):

```go
// Read ...
//
// Returns [os.ErrDeadlineExceeded] if a deadline was set and the perf ring buffer
// was empty. Otherwise returns a record and no error, even if the deadline was
// exceeded.
```

This means we will never return ErrDeadlineExceeded as long as there is data in the rings. Kind of crucial for the purpose of the change, but also weird? Not sure.

Reply:
Thinking more about this, should we continue with this draft?
It is a slightly change in the API (behaviour) but most users should benefit, I think.

Reply:
How are we going to solve this? Do you think this behaviour is OK?

> This means we will never return ErrDeadlineExceeded as long as there is data in the rings.

Reply:
Yeah - thats the API (behaviour) change im refering to. Without a bigger change in the API, I think it is not possible to solve this.
But I also think, this behaivour is ok and should benefit most users.